iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/unique_port_id.hpp

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_POSH_POPO_BUILDING_BLOCKS_UNIQUE_PORT_ID_HPP
+#define IOX_POSH_POPO_BUILDING_BLOCKS_UNIQUE_PORT_ID_HPP
+
+#include "iceoryx_hoofs/cxx/newtype.hpp"
+#include "iceoryx_hoofs/error_handling/error_handling.hpp"
+
+#include <atomic>
+#include <cstdint>
+#include <limits>
+
+namespace iox
+{
+namespace popo
+{
+/// @brief Struct to signal the constructor to create an invalid id
+struct InvalidPortId_t
+{
+};
+constexpr InvalidPortId_t InvalidPortId = InvalidPortId_t();
+
+/// @brief A counter which is monotonically advancing with each newly created instance of UniquePortId. Additionally
+///        it contains a unique RouDi id to be able to differentiate the sample sources in a multi publisher multi
+///        subscriber pattern where samples are exchanged over network via a third party middleware.
+///        The unique RouDi id must be set manually when RouDi is started and it must be ensured to be unique for a
+///        given instance for this feature to be used to its full extend.
+class UniquePortId : public cxx::NewType<uint64_t,
+                                         cxx::newtype::ProtectedConstructByValueCopy,
+                                         cxx::newtype::Comparable,
+                                         cxx::newtype::Sortable,
+                                         cxx::newtype::Convertable,
+                                         cxx::newtype::CopyConstructable,
+                                         cxx::newtype::MoveConstructable,
+                                         cxx::newtype::CopyAssignable,
+                                         cxx::newtype::MoveAssignable>
+{
+  public:
+    using ThisType::ThisType;
+
+    /// @brief The constructor creates an id which is greater than the
+    ///         previous created id
+    UniquePortId() noexcept;
+
+    /// @brief Constructor which creates an invalid id
```

Comment:
This again smells somehow like a nullable type. Do we really need this or is it possible to pack the `UniquePortId` into an `optional` when it is not valid.
What is the use case for having an invalid `UniquePortId`?

We even could get rid of `isValid` and with that I am sure maybe even of some error handling or additional tests?!

Reply:
@elfenpiff I agree with you, but if you don't mind, I would create a separate issue for this. The PR is already quite big and this change would impact many other files. Are you okay with doing it in a follow up PR?

Reply:
After some thinking a came to the conclusion that we might not want to use an optional here. The `UniquePortId` is used in the `ChunkHeader` and we must be very careful when member of this class changes and always increase the version number. There could be recorded data which needs to be replayed and subtle bugs might occur when the version match but the binary layout and semantics change. With an complex type like an `optional` it might be possible to accidentally change the binary layout of the `ChunkHeader` without noticing it by just refactoring the `optional`.

In this special case, I would keep an invalid `UniquePortId`. To not break source compatibility I would also keep the `UniquePortId` instead of using the underlying type but I would also add some tests to check to notice when the underlying type changes or the `UniquePortId` gains new members.